Fix issue 16479: No namespace substitution for C++ mangling on POSIX

Geod24 opened the pull request:
```
The C++ ABI used by POSIX is the Itanium C++ ABI.
Reference document available here: https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling

One important (and tricky) part of the ABI is the substitutions being done,
in order to reduce the bloat introduced by long symbol names,
a typical issue when using templates heavily (of which D was not exempt).

There are 2 kinds of substitutions: component substitution and template parameter substitution.
Component substitution replaces repeated parts of the symbol with `S[X]_`,
template parameter substitution replaces occurences of template parameters with `T[X]_`.
`X` represents a base36 index into the array of components or
template parameters already encountered so far, respectively.

This substitution is done on an identity basis, which means that the templated function
`template<typename T> int foo()` instantiated with `int` will be mangled as `_Z3fooIiE*i*v`
(asterisks are emphasis and not part of the mangling) while it would be mangled as `_Z3fooIiE*T_*v`
if the definition was `template<typename T> T foo()`.

Moreover, experience with C++ compilers shows that component substitution is prefered over
template parameter substitution, such as `template<typename T> T foo(T)` is mangled as
`_Z3fooIiET_S0_` when instantiated with `int` and not `_Z3fooIiET_T_` as would be the case
if template substitution was prefered.

This is just brushing the surface of the problem, since only template type parameters have been
mentioned so far, but other kind (aliases, values) are also concerned.
Substitution also needs to happen if a template parameter is part of another type,
such as the `template<typename T> array<T>* foo (T, int)`, which, when instantiated with `int`,
is mangled as `_Z3fooIiEP5arrayIT_ES1_i`.

For more detailed test cases, see `test/compilable/cppmangle.d`.

The main issue encountered while implementing this in DMD is that there's no easy way to know
if a type (which is part of the function's type, e.g. parameters and return value)
was a template parameter or not, as DMD merges types, so in the previously mentioned
`template<typename T> int foo()` vs `template<typename T> T foo()` the template instantiation
will come with the same exact two pointer to the singleton `int` type.

Moreover, DMD does destructive semantic analysis, meaning that objects gets mutated,
pointers get replaced, aliases get resolved, and information gets lost.

After different approaches where taken, the most practical and reliable approach devised was to
provide a `visit` overload for non-resolved AST type `TypeIdentifier` and `TypeInstance`,
and compare the identifier to that of the template definition.
Fallback to post-semantic type when it isn't found.

Note that no attempt has been made whatsoever to handle the mess that would result from
expressions themselves being mangled. The reference doc for the ABI mentions that
"[...] this mangling is quite similar to the source token stream. (C++ Standard reference 14.5.5.1p5.)".
Original quote:
https://itanium-cxx-abi.github.io/cxx-abi/abi.html#expressions (5.1.6 Expressions)
```

~Please don't merge yet, I want to attempt to port some std types to D and see if I'm missing something. But I think this is ready for reviews.~

Geod24 committed:
Fix issue 16479: No namespace substitution for C++ mangling on POSIX

The C++ ABI used by POSIX is the Itanium C++ ABI.
Reference document available here: https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling

One important (and tricky) part of the ABI is the substitutions being done,
in order to reduce the bloat introduced by long symbol names,
a typical issue when using templates heavily (of which D was not exempt).

There are 2 kinds of substitutions: component substitution and template parameter substitution.
Component substitution replaces repeated parts of the symbol with `S[X]_`,
template parameter substitution replaces occurences of template parameters with `T[X]_`.
`X` represents a base36 index into the array of components or
template parameters already encountered so far, respectively.

This substitution is done on an identity basis, which means that the templated function
`template<typename T> int foo()` instantiated with `int` will be mangled as `_Z3fooIiE*i*v`
(asterisks are emphasis and not part of the mangling) while it would be mangled as `_Z3fooIiE*T_*v`
if the definition was `template<typename T> T foo()`.

Moreover, experience with C++ compilers shows that component substitution is prefered over
template parameter substitution, such as `template<typename T> T foo(T)` is mangled as
`_Z3fooIiET_*S0_*` when instantiated with `int` and not `_Z3fooIiET_*T_*` as would be the case
if template substitution was prefered.

This is just brushing the surface of the problem, since only template type parameters have been
mentioned so far, but other kind (aliases, values) are also concerned.
Substitution also needs to happen if a template parameter is part of another type,
such as the `template<typename T> array<T>* foo (T, int)`, which, when instantiated with `int`,
is mangled as `_Z3fooIiEP5arrayIT_ES1_i`.

For more detailed test cases, see `test/compilable/cppmangle.d`.

The main issue encountered while implementing this in DMD is that there's no easy way to know
if a type (which is part of the function's type, e.g. parameters and return value)
was a template parameter or not, as DMD merges types, so in the previously mentioned
`template<typename T> int foo()` vs `template<typename T> T foo()` the template instantiation
will come with the same exact two pointer to the singleton `int` type.

Moreover, DMD does destructive semantic analysis, meaning that objects gets mutated,
pointers get replaced, aliases get resolved, and information gets lost.

After different approaches where taken, the most practical and reliable approach devised was to
provide a `visit` overload for non-resolved AST type `TypeIdentifier` and `TypeInstance`,
and compare the identifier to that of the template definition.
Fallback to post-semantic type when it isn't found.

Note that no attempt has been made whatsoever to handle the mess that would result from
expressions themselves being mangled. The reference doc for the ABI mentions that
"[...] this mangling is quite similar to the source token stream. (C++ Standard reference 14.5.5.1p5.)".
Original quote:
https://itanium-cxx-abi.github.io/cxx-abi/abi.html#expressions (5.1.6 Expressions)

Geod24 committed:
Turn component comparison to a visitor

The issue with the current scheme of comparing object is that it's hard to match C++'s semantics.
The first issue uncovered was with Namespace, which need to compare equals if they have the same name,
even if they are from 2 different modules.
Since the same behaviour needs to happen for template parameter substitution,
many more types need to be handled, and most of them have a common implementation,
which is a good indicator we should use a Visitor.

Geod24 commented:
Updated, with loads of tests
@TurkeyMan it might solve some (and hopefully all) of your issues.

I welcome other tests cases, but one of my main target was:
```cpp
#include <vector>

template<typename T>
std::vector<T>* foo() { return 0; }

int main ()
{
    foo<int>();
    return 0;
}
```

```console
$ nm a.out
0000000100000f90 T __Z3fooIiEPNSt3__16vectorIT_NS0_9allocatorIS2_EEEEv
0000000100000000 T __mh_execute_header
0000000100000f60 T _main
                 U dyld_stub_binder
```

```D
extern(C++, `std`) {
    extern (C++, `__1`) {
        struct allocator (T);
        struct vector(T, alloc = allocator!T);
    }
}

extern(C++) vector!(T)* foo(T)();

void main ()
{
    assert(foo!int() is null);
}
```

```console
$ nm abo.o
0000000000000050 s EH_frame0
00000000000000b8 s _TMP0
0000000000000040 S __D3abo12__ModuleInfoZ
0000000000000000 T __Dmain
                 U __Dmain
0000000000000068 S __Dmain.eh
                 U __Z3fooIiEPNSt3__16vectorIT_NS0_9allocatorIS2_EEEEv # Yay
                 U __d_assertp
                 U __d_run_main
                 U _main
0000000000000028 T _main
0000000000000090 S _main.eh
```

jacob-carlborg commented on src/dmd/cppmangle.d:
```diff
@@ -101,8 +101,52 @@ bool isPrimaryDtor(const Dsymbol sym)
 private final class CppMangleVisitor : Visitor
 {
+    /// Context used when processing pre-semantic AST
+    private Context ctx;
```
Can we please avoid abbreviations like this.

Geod24 replied:
I wish you could convince Walter, because `ctx` is much more widespread than the various 2-letters type abbreviations which are so common in DMD. Anyway, I'll update with the full name.

jacob-carlborg replied:
A project wide search for `ctx` gives back 25 results and for `context` it gives back 253 results. This is just a plain text search which includes comments and everything.

jacob-carlborg commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1426,391 @@ extern(C++):
+        t.accept(this_);
+    else if (Tuple t = isTuple(o))
+        this_.visit(t);
+    else {
```
A bit inconsistent here that none of the other branches have curly braces.

Geod24 commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1426,391 @@ extern(C++):
+     *   ns = C++ namespace to do substitution for
+     *
+     * Returns:
+     *  Index of the entry, if found, or `-1` otherwise
```
That's a copy-pasta leftover, removed

Geod24 commented:
Updated according to @jacob-carlborg 's suggestions

jacob-carlborg commented:
> Updated according to jacob-carlborg 's suggestions

I don't think I can be of any more help than for these stylistics issues.

Geod24 commented:
Added some tests to instantiate some stdlib containers and return them (`test/runnable/cpp_stdlib.d`).
It fails on Linux because files need to be compiled with `-std=c++11` and on OSX32 because of `size_t` mangling, but it does pass on OSX64 and FreeBSD.

Geod24 committed:
Testsuite: Let user provide extra argument for C++ source

We need to compile some code with `-std=c++11` but not all the tests.

Geod24 committed:
cppmangle: Split special mangling of std symbol in its own function

Because it's going to be reused later.
Also switch from `s.ident` to `ti.name`.

Geod24 committed:
Testsuite: Don't attempt to compile extra_sources on disabled tests

Geod24 commented:
Should be ready. Extra test (cpp_stdlib) disabled on osx32 and windows.

TurkeyMan commented:
LGTM, but it's a truly gigantic patch!
The tests appear to prove it works though.

Geod24 commented:
Okay, so this is green and IMO ready, but I don't know who can review / approve / merge this.

The last (and main) commit is indeed gigantic, but I don't see a way to break it down further that would make sense. Fortunately it's backed by an unusual amount of tests, which hopefully will provide some confidence to the poor reviewer(s).

Last but not least, most of the code in that commit should only trigger when templates are used. Given anything template-related is currently broken, even if this PR doesn't manage to capture all the cases, it will already be a big win over something which is currently completely broken.

I'd very much like this to get merged sooner than later, so we can focus on druntime's `stdcpp` bindings and other related issues (e.g. `classinfo`).

ibuclaw commented:
Can you wait for the weekend? I'll probably have some time then.

Geod24 commented:
Sure can. Thanks!

Geod24 commented:
Ping @ibuclaw

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+     * This is because there is no easy way to link a type to the template
+     * parameters once semantic ran, because:
+     * - the `TemplateInstance` installs aliases in its scope to its params
+     * - `AliasDeclaration` are resolved in many places, so we cannot keep a link
```
I think you mean "`AliasDeclaration`s" as a plural here.

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+
+    /**
+     * Helper function to go through the `TemplateParameter`s
+     * and perform a substitution, if possible.
```
What column length are you keeping yourself to here?  I would say 80 is standard, anything shorter just looks out of place.

Geod24 replied:
Yep 80. Some places have seen a few iterations and might look weird. I'm also not sure what the correct rules for breaking a sentence are in English. Fixed this.

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+     *   The `TemplateParameter` that was used, or `null`.
+     */
+    private bool writeTemplateSubstitution (
+        const ref Identifier ident, TemplateParameters* params, Type type)
```
Keeping parameters on the same line is more conformant to coding style.

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+     * and perform a substitution, if possible.
+     *
+     * Returns:
+     *   The `TemplateParameter` that was used, or `null`.
```
This comment - `TemplateParameter` or `null` - does not match the return type `bool`.

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+
+                // Since we didn't run semantic yet, all we get are
+                // `TypeIdentifier`, but we need to check if it's an expression
+                // or whatnot
```
This (and maybe other places, I haven't looked too deep), could be rephrased to use a little less `we`.

For instance (feel free to correct as you see fit):
```
Since semantic hasn't run yet, all that is handled here are `TypeIdentifier`,
but we still need to check if it's an expression or not.
```

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+    }
+
+    /**
+     * Given a template instance `t`, write it's qualified name
```
`its`

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+     *       to mangle (one level deep)
+     *
+     * Returns:
+     *   `true` if the name was qualified and need an ending `E`
```
and needs an ending `E`.

You could also say requires here, need sounds urgent. :-)

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1373,4 +1453,400 @@ extern(C++):
+        {
+            if (isStd(sym2))
+            {
+                bool useless;
```
If only used `bool* = null` instead of an `out bool`...

Perhaps `unused` is a better name than `useless` for a local var.

ibuclaw reviewed (changes requested):
Few nits, nothing that I feel is preventing merge here.

Geod24 commented on src/dmd/cppmangle.d:
```diff
@@ -606,6 +611,65 @@ private final class CppMangleVisitor : Visitor
     }
 
+    /**
+     * Write common substitution for standard types, such as std::allocator
```
AFAIK, we don't use full stop for the title as it's supposed to be a one-liner.
It's however used in the description, which can be any number of sentences.

Geod24 commented on src/dmd/cppmangle.d:
```diff
@@ -101,8 +101,52 @@ bool isPrimaryDtor(const Dsymbol sym)
     return dtor == ad.primaryDtor;
 }
 
+/// Context used when processing pre-semantic AST
```
You mean full stop ? If so see my previous comment.

ibuclaw replied:
I'll highlight another example then.

Geod24 commented on test/compilable/cppmangle.d:
```diff
@@ -771,3 +773,165 @@ version(Windows)
+    // Make sure aliases are still resolved
+    alias Alias16479 = int;
+    Alias16479 func16479_0 (FuncT1) (FuncT1, Alias16479);
+    static assert(func16479_0!(int).mangleof == `_Z11func16479_0IiEiT_i`);
```
Agreed, I originally went with full mangleof testing (makes development easier), but then later realized I was missing some things, which is why the link testing was added.
I'll keep it in mind in the future, and maybe will move some test cases over in a later iteration.

Geod24 commented on test/runnable/cpp_stdlib.d:
```diff
@@ -0,0 +1,59 @@
+// DISABLED: win32 win64 osx32
+// EXTRA_CPP_SOURCES: cpp_stdlib.cpp
+// CXXFLAGS: -std=c++11
```
Yeah, I originally went full C++11 but `throwle` failed to link.
Do you have any document about the mangling changes between 98 and 11 ?
I didn't find anything relevant and figured this was too big of a matter to address in this PR.

Geod24 commented on src/dmd/cppmangle.d:
```diff
@@ -101,8 +101,52 @@ bool isPrimaryDtor(const Dsymbol sym)
+    /**
+     * Reset the context to a previous one, making any adjustment necessary
+     */
+    private void reset (ref Context prev)
```
Changed to `push` / `pop`. Internal chaining sounds nice, but the current pattern is IMO simple enough and allocation-less that it does not provide much of an improvement.

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -101,8 +101,52 @@ bool isPrimaryDtor(const Dsymbol sym)
+     * Params:
+     *   next = Value to set `this.res` to
+     *          If `this.res` is `null`, the expression is not evalutated
+     *          This allow this code to be used even when no context is needed.
```
As the first line didn't have a period, when reading this I started the parsing the second line as part of the same sentence.

Not disputing whether it looks fine in the generated docs, however I won't likely ever read from the web. :-)

ibuclaw commented on src/dmd/cppmangle.d:
```diff
@@ -1127,11 +1209,6 @@ extern(C++):
-    override void visit(Type t)
-    {
-        error(t);
-    }
```
It seemed to be already appropriately grouped with other type visit methods here, which is why I asked.

Geod24 replied:
Oups, that's right. It was moved because I ordered them from most specialized to least specialized.
Can revert if you want.

ibuclaw replied:
The less line changes you can fit this in the better. I flagged this because initially I thought you removed it outright.

Geod24 commented:
Updated:
```diff
diff --git a/src/dmd/cppmangle.d b/src/dmd/cppmangle.d
index 2083625fe..b5647e94c 100644
--- a/src/dmd/cppmangle.d
+++ b/src/dmd/cppmangle.d
@@ -118,8 +118,8 @@ private struct Context
      * Helper function to track `res`
      *
      * Params:
-     *   next = Value to set `this.res` to
-     *          If `this.res` is `null`, the expression is not evalutated
+     *   next = Value to set `this.res` to.
+     *          If `this.res` is `null`, the expression is not evalutated.
      *          This allow this code to be used even when no context is needed.
      *
      * Returns:
```

Geod24 commented:
Anything else @ibuclaw ?

ibuclaw commented:
@Geod24 please read through all comments and documentation you've written up - preferably out loud. If you don't pick up any mistakes, then it's fine.

thewilsonator commented on test/compilable/cppmangle.d:
```diff
@@ -771,3 +773,165 @@ version(Windows)
+
+    // Now let's have a bit of fun with alias parameters,
+    // starting with C functions
+    // TODO: Why is this mangled by g++:
```
Is this working?

Geod24 replied:
Re-looking at this, it seems correct.
I'll do what Iain requested in a few hours and will update the comment.

thewilsonator reviewed (approved):
Otherwise LGTM, waiting on DAutoTest to go green.

Geod24 commented:
> @Geod24 please read through all comments and documentation you've written up - preferably out loud. If you don't pick up any mistakes, then it's fine.

Thanks, did pick up quite a few. Removed historical artifacts, some typos, and some usage of "we".

thewilsonator commented:
@ibuclaw merge this?
